library/src/main/java/com/chuckerteam/chucker/api/ChuckerInterceptor.kt

```diff
@@ -140,50 +160,62 @@ class ChuckerInterceptor @JvmOverloads constructor(
 
             tookMs = (response.receivedResponseAtMillis() - response.sentRequestAtMillis())
         }
-
-        return if (responseEncodingIsSupported) {
-            processResponseBody(response, transaction)
-        } else {
-            response
-        }
     }
 
     /**
-     * Processes a [ResponseBody] and populates corresponding fields of a [HttpTransaction].
+     * Multi casts a [Response] body if it is available and downstreams it to a file which will
+     * be available in [onResponseBodyReady] callback once end user consumes the body.
      */
-    private fun processResponseBody(response: Response, transaction: HttpTransaction): Response {
-        val responseBody = response.body() ?: return response
+    private fun multiCastResponseBody(
+        response: Response,
+        onResponseBodyReady: (File?) -> Unit
+    ): Response {
+        val responseBody = response.body()
+        if (responseBody == null) {
+            onResponseBodyReady(null)
+            return response
+        }
 
         val contentType = responseBody.contentType()
-        val charset = contentType?.charset(UTF8) ?: UTF8
         val contentLength = responseBody.contentLength()
 
-        val responseSource = if (response.isGzipped) {
-            GzipSource(responseBody.source())
-        } else {
-            responseBody.source()
-        }
-        val buffer = Buffer().apply { responseSource.use { writeAll(it) } }
+        val teeSource = TeeSource(
+            responseBody.source(),
+            fileFactory.create(),
+            maxContentLength,
+            onResponseBodyReady
+        )
+
+        return response.newBuilder()
+            .body(ResponseBody.create(contentType, contentLength, Okio.buffer(teeSource)))
```

Comment:
I insist on returning the original `Response` object further in chain without any builders, cloning, etc.
We had enough side effects and issues with providing processed or somehow changed objects. 
And while I see that here we should get the same object I still don't like that it is recreated. 
In some issues we already mentioned that Chucker should be as transparent as possible and this is possible only if we do copy only for Chucker processing without anything like that.

Reply:
I don't see how it would be possible to stream original bytes to two sinks at the same time without replacing the response. This is a mechanism that is used under the hood by `OkHttp` for caching (just a simpler version). It is also recommended way by Square (https://github.com/square/okio/issues/186). Do you have any other approach in mind that would solve the problems addressed by this?